pkg/kubernetes/client/client.go

```diff
@@ -0,0 +1,201 @@
+		WithLabels(getDefaultLabels(name)).
+		WithSpec(jv.JivaVolumeSpec{
+			PV:       name,
+			Capacity: req.GetCapacityRange().GetRequiredBytes(),
+			ReplicaSC: func(sc string) string {
+				if sc == "" {
+					return "openebs-hostpath"
+				}
+				return sc
+			}(sc),
+			ReplicaResource: func(req *csi.CreateVolumeRequest) v1.ResourceRequirements {
+				return v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse(jivavolume.HasResourceParameters(req)("replicaMinCPU")),
+						v1.ResourceMemory: resource.MustParse(jivavolume.HasResourceParameters(req)("replicaMinMemory")),
+					},
+					Limits: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse(jivavolume.HasResourceParameters(req)("replicaMaxCPU")),
+						v1.ResourceMemory: resource.MustParse(jivavolume.HasResourceParameters(req)("replicaMaxMemory")),
+					},
+				}
+			}(req),
+
+			TargetResource: func(req *csi.CreateVolumeRequest) v1.ResourceRequirements {
+				return v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse(jivavolume.HasResourceParameters(req)("targetMinCPU")),
+						v1.ResourceMemory: resource.MustParse(jivavolume.HasResourceParameters(req)("targetMinMemory")),
+					},
+					Limits: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse(jivavolume.HasResourceParameters(req)("targetMaxCPU")),
+						v1.ResourceMemory: resource.MustParse(jivavolume.HasResourceParameters(req)("targetMaxMemory")),
+					},
+				}
+			}(req),
+			ReplicationFactor: req.GetParameters()["replicaCount"],
+		})
+
+	if jiva.Errs != nil {
+		return fmt.Errorf("failed to build JivaVolume CR, err: %v", jiva.Errs)
+	}
+
+	obj := jiva.Instance()
+	err := cl.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: ns}, obj)
+	if err != nil && errors.IsNotFound(err) {
+		logrus.Infof("Creating a new JivaVolume CR {name: %v, namespace: %v}", name, ns)
+		err = cl.client.Create(context.TODO(), obj)
+		if err != nil {
+			return fmt.Errorf("failed to create JivaVolume CR, err: %v", err)
+		}
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to get the JivaVolume details, err: %v", err)
+	}
+	return nil
+}
+
+// DeleteJivaVolume delete the JivaVolume CR
+func (cl *Client) DeleteJivaVolume(req *csi.DeleteVolumeRequest) error {
+	volumeID := req.GetVolumeId()
+	obj := &jv.JivaVolumeList{}
+	opts := []client.ListOption{
+		client.MatchingLabels(getDefaultLabels(volumeID)),
+	}
+	err := cl.client.List(context.TODO(), obj, opts...)
+	//	err := cl.client.Get(context.TODO(), types.NamespacedName{Name: volumeID, Namespace: "openebs"}, obj)
+	if err != nil && errors.IsNotFound(err) {
+		logrus.Warningf("DeleteVolume: failed to list JivaVolume CR: {%v}, err: %v, ignore deletion...", volumeID, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	logrus.Debugf("DeleteVolume: object: {%+v}", obj)
+	instance := obj.Items[0].DeepCopy()
```

Comment:
No, since we donot have namespace info, it can be deployed into any namespace, so listing is the only option we will have. I will add an nil check over here.